### PR TITLE
Fix dropped ingress packets in azure

### PR DESCRIPTION
**Pull request by vomba**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [ ] personal data beyond what is necessary for interacting with this pull request, nor
> - [ ] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [x] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

This PR fixes the problem with dropped ingress packets to the ingress nginx on azure, it also enables preserving the source IP of incoming traffic.

- Fixes https://github.com/elastisys/ck8s-issue-tracker/issues/250

#### Checklist

- [ ] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [x] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests
- Config checks:
  - [ ] The schema was updated

### CODE REVIEW

**Comment by vomba** [comment]
- For `whitelist-source-range` it was there in the `common-config` by default, i don't have a clear idea if it is needed or not.
- The network policies that does look very weird, did that happen after you did the init? i don't remember touching anything there to make it behave like that. running a `diff` doesn't show that.
- `allocateLoadBalancerNodePorts` to use direct routing from load balancer and fix the NetworkPolicies issue, which gave the opportunity to run `externalTrafficPolicy` and preserve the source IP

File: bin/init.bash [hunk 1]
```diff
@@ -295,8 +300,16 @@ set_nginx_config() {
     replace_set_me "${file}" '.ingressNginx.controller.service.enabled' "${service_enabled}"
     replace_set_me "${file}" '.networkPolicies.global.externalLoadBalancer' "${external_load_balancer}"
     replace_set_me "${file}" '.networkPolicies.global.ingressUsingHostNetwork' "${ingress_using_host_network}"
-    replace_set_me "${file}" '.ingressNginx.controller.service.allocateLoadBalancerNodePorts' 'true'
+    replace_set_me "${file}" '.ingressNginx.controller.service.allocateLoadBalancerNodePorts' "${allocate_loadbalancer_node_ports}"
 
+    if [ "${CK8S_CLOUD_PROVIDER}" = "azure" ]; then
+        replace_set_me "${file}" '.externalTrafficPolicy.local' 'true'
+        replace_set_me "${file}" '.networkPolicies.ingressNginx.ingressOverride.enabled' 'true'
+        replace_set_me "${file}" '.networkPolicies.ingressNginx.ingressOverride.ips' "[ 0.0.0.0/0 ]"
```
Comment (simonklb) [inline bin/init.bash:308]: The network policy bug is that this is currently being replaced in common but it should be replaced in the sc and wc configs

**Comment by simonklb** [comment]
> * `allocateLoadBalancerNodePorts` to use direct routing from load balancer and fix the NetworkPolicies issue

Are you sure this is supported by Azure? I think it uses node ports for the health probes.

**Comment by simonklb** [comment]
Note that disabling node port allocation does not also automatically remove the node ports: https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation

This means that even though you set it to false and applies it on an existing cluster it will not take effect.

**Comment by vomba** [comment]
> > * `allocateLoadBalancerNodePorts` to use direct routing from load balancer and fix the NetworkPolicies issue
> 
> Are you sure this is supported by Azure? I think it uses node ports for the health probes.

It creates a specific port for health checks.

**Comment by vomba** [comment]
> Note that disabling node port allocation does not also automatically remove the node ports: https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-nodeport-allocation
> 
> This means that even though you set it to false and applies it on an existing cluster it will not take effect.

Yeah, i completely removed and then reapplied ingress-nginx so that the new load balancers get created without them.

**Review by Xartos — APPROVED** [review]

File: bin/init.bash (hunk 1 above)
Comment (simonklb) [reply bin/init.bash:308]: Yes, but it's not a `set-me` value in that config file.

Comment (vomba) [reply bin/init.bash:308]: Weirdly enough, when i add `ips: set-me` i get bad expression error, but without it, the init works fine.

File: config/config/common-config.yaml [hunk 2]
```diff
@@ -1041,7 +1041,8 @@ networkPolicies:
   ingressNginx:
     enabled: true
     ingressOverride:
-      enabled : false
+      enabled : set-me
```
Comment (aarnq) [inline config/config/common-config.yaml:1044]:
```suggestion
      enabled: set-me
```
Nit

**Review by aarnq — APPROVED** [review]

File: bin/init.bash (hunk 1 above)
Comment (simonklb) [reply bin/init.bash:308]: Sent a fix for this here: https://github.com/elastisys/compliantkubernetes-apps/pull/2095

**Pull request merged**